packages/flame/lib/src/components/clip_component.dart

```diff
@@ -0,0 +1,132 @@
+import 'dart:ui';
+
+import 'package:flame/components.dart';
+import 'package:flame/experimental.dart';
+
+/// {@template clip_component}
+/// A component that will clip its content.
+/// {@endtemplate}
+abstract class ClipComponent extends PositionComponent {
+  /// {@macro clip_component}
+  ///
+  /// Clips the canvas based its shape and size.
+  ClipComponent({
+    super.position,
+    super.size,
+    super.scale,
+    super.angle,
+    super.anchor,
+    super.children,
+    super.priority,
+  });
+
+  late Path _path;
+  late Shape _shape;
+
+  @override
+  Future<void> onLoad() async {
+    _prepare();
+    size.addListener(_prepare);
+  }
+
+  Shape _buildShape();
+
+  void _prepare() {
+    _shape = _buildShape();
+    _path = _shape.asPath();
+  }
+
+  @override
+  void render(Canvas canvas) => canvas.clipPath(_path);
+
+  @override
+  bool containsPoint(Vector2 point) {
+    return _shape.containsPoint(point - position);
+  }
+
+  @override
+  bool containsLocalPoint(Vector2 point) {
+    return _shape.containsPoint(point);
+  }
+}
+
+/// {@template circle_clip_component}
+///
+/// A component that will clip its content to a circle.
+class CircleClipComponent extends ClipComponent {
+  /// {@macro circle_clip_component}
+  ///
+  /// Clips the canvas in the form of a circle based on its size.
+  CircleClipComponent({
+    super.position,
+    super.size,
+    super.scale,
+    super.angle,
+    super.anchor,
+    super.children,
+    super.priority,
+  });
+
+  @override
+  Shape _buildShape() {
```

Comment:
Why is this preferred over having the shape be passed to the ClipComponent constructor?

Reply:
We need to recreate the shape based on size of the component, otherwise the user would need to pass the Shape and the Size, and this would be confusing.

Reply:
Would a parameter named shapeBuilder that receives a Size suffice?

Reply:
A shape builder is not a bad idea, but I would think of it as a way of giving more freedom to the user to define their own way of building the shapes. But I would create `.builder` constructor to it and keep the other constructors as requiring the user to define the shape builder always would not be good API IMO.

Reply:
A .builder constructor is very interesting. I like where this ja heading.

Out of curiosity, under what circumstances do you think the shapeBuilder will fail to be a "good" API?

Reply:
Imagine requiring the user to write this every time they want a circle:

```dart
ClipComponent(builder: (size) => ..., size: Vector2.all(50), ...);
```

When they could instead do:

```dart
ClipComponent.circle(size: Vector2.all(50), ...);
```

But with a `builder` factory, we can have the best of both things.

Reply:
The above is exactly what I was referring to and would like. Are we happy with it?

So the API would be:
```dart
ClipComponent(builder: (size) => ..., size: Vector2.all(50), ...);
ClipComponent.circle(size: Vector2.all(50), ...);
ClipComponent.polygon(size: Vector2.all(50), ...);
...
```

Reply:
Yep, that is what I will implement.

Reply:
Looking forward to it! Thanks @erickzanardo for the work put into this!

Reply:
Updated! 🚀